Add Chips post (#2127)
GoogleChrome/developer.chrome.com: site/en/docs/privacy-sandbox/chips/index.md

---
layout: 'layouts/doc-post.njk'
title: 'CHIPS—Cookies Having Independent Partitioned State'
subhead: >
  Allow developers to opt-in a cookie to "partitioned" storage, with a separate cookie jar per top-level site.
description: >
  CHIPS is a Privacy Sandbox proposal that will allow cookies to be set by a third-party service, but only read within the context of the top-level site where they were initially set
date: 2022-02-15
authors:
  - mihajlija
---

> 

Author: OT coming soon, will move that section up then. Right now we don't have much info to share so it can't really follow the template.

For example, when a user visits site A, embedded content from site C can set a cookie on the user's machine in response to the cross-site request. If the user then visits site B which also embeds C, site C could access the same cookie that has been set earlier when the user visited site A. This allows site C to compile a user's browsing activity across site A, site B, and every site that it is embedded on.

To protect user privacy, browser vendors are placing restrictions on this behavior and phasing out support for third-party cookies.

While cross-site tracking is an issue, there are still valid cross-site cookie use cases on the web today which can be achieved in a privacy-preserving way with cookie partitioning.

### Use cases

For example, the site `retail.example` may want to work with a third-party service `support.chat.example` to embed a support chat box on its site. Many embeddable chat services today rely on cookies to save interaction history. 

{% Img src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/jsxgCpkMRXwXughPjg7j.png", alt="ALT_TEXT_HERE", width="800", height="620" %}

> Without the ability to set a cross-site cookie, `support.chat.example` could instead rely on `retail.example` passing along their first-party session identifier (or some derived value of it). In that case, every website that `support.chat.example` is embedded on would require additional setup to pass along the state.  

Reviewer: I'm actually not sure what state is -- and maybe that's my lack of knowledge of cookies. 

I reviewed the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies). It's possible I'm overthinking it with my lack of knowledge, but it may be helpful to define. Such as:

> In that case, additional setup would be required to pass along session information and personalization (otherwise known as the _state_) for every website that embeds `support.chat.example`.

Alternatively, `support.chat.example` may request `retail.example` to embed JavaScript that's hosted on `support.chat.example` on `retail.example` pages. This introduces security risks, because it allows the `support.chat.example` script to have elevated privileges on `retail.example`, such as the ability to access authentication cookies.

Example use cases for CHIPS include any scenarios where cross-site subresources require some notion of session or persistent state that is scoped to a user's activity on a single top-level site, such as:

+   Third-party chat embeds
+   Third-party map embeds
+   Subresource CDN load balancing
+   Headless CMS providers
+   Sandbox domains for serving untrusted user content (such as googleusercontent.com and githubusercontent.com)

> +   Third-party CDNs that use cookies to serve content that's access-controlled based on the authentication status on the first-party site (for example, profile pictures on social media sites hosted on third-party CDNs)

Reviewer: ```suggestion
+   Third-party CDNs that use cookies to serve content that's access-controlled by the authentication status on the first-party site (for example, profile pictures on social media sites hosted on third-party CDNs)
```

> +   Front-end frameworks that rely on remote APIs using cookies on their requests

Reviewer: Do you have an example of this that could be shared?

Author: The example is in the parenthesis. Did you miss it or is it not clear enough?

### How does it work?

> CHIPS introduces a new cookie attribute, `Partitioned`, to support cross-site cookies that are partitioned by top-level context.  

Reviewer: Consider replacing "partitioned by top-level context" with "divided by top-level context." Since partitioned has been used frequently and not defined in another way, it may be helpful for users to have another way of thinking about that language.

> Under this proposal, when a user visits site A and embedded content from site C sets a cookie with the `Partitioned `attribute, the cookie is saved in a partitioned jar designated only for cookies that the site C sets when it's embedded on site A. The browser would only send that cookie when the top-level site is A.   

Reviewer: This is a long, complex sentence. I'm not *quite* sure what to recommend, so I don't have suggestions yet--but it's a lot of attributes in consideration which are cognitively significant. Perhaps it can be broken into two sentences.

Reviewer: ```suggestion
From our example, a user visits site A which embeds content from site C. Site C sets a cookie with the `Partitioned` attribute. Under this proposal, when a user visits site A, site C's cookie is saved in a partitioned jar designated only for cookies set by site C when it's embedded on site A. The browser will only send the cookie when the top-level site is A.  
```

Reviewer: I'm not sure that the rewrite does what you need. Just a first offering.

When the user visits a new site, for example site B, an embedded C frame would not receive the cookie that was set when C was embedded in site A. 

{% Img src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/PODInBIXZrgGeUoFhipj.png", alt="ALT_TEXT_HERE", width="800", height="443" %}

If a user visits site C as a top level website, the partitioned cookie that C set when it was embedded in A will not be sent in that request either.

{% Img src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/FaQYaZyTsAxCm8GvvVc8.png", alt="ALT_TEXT_HERE", width="800", height="608" %}

### Why it's important to opt into cookie partitioning

As browsers are phasing out unpartitioned third-party cookies, a couple of other approaches to partitioning have been attempted.

Firefox announced that they are [partitioning all third-party cookies by default](https://hacks.mozilla.org/2021/02/introducing-state-partitioning/) in their ETP Strict mode, so all cross-site cookies are partitioned by the top-level site.   

> However, partitioning cookies without a third-party opt-in can lead to unexpected bugs, since some third party services have built servers with the expectation of an unpartitioned third-party cookie.

Reviewer: ```suggestion
However, partitioning cookies without a third-party opt-in can lead to unexpected bugs, since some third-party services have built servers which expect an unpartitioned third-party cookie.
```

[Safari previously tried partitioning cookies based on heuristics](https://webkit.org/blog/8613/intelligent-tracking-prevention-2-1/), but eventually chose to block them altogether, citing developer confusion as one of the reasons.

What sets CHIPS apart from existing implementations of partitioned cookies is the third-party opt-in. Cookies must be set with a new attribute in order to be sent on cross-party requests once (unpartitioned) third-party cookies are obsoleted.

While third-party cookies still exist, the `Partitioned` attribute provides an opt-in to a more restrictive, more secure type of cookie behavior. CHIPS is an important step to help services make a smooth transition to a future without third-party cookies.

### CHIPS design details

#### Partitioning model

Today, cookies are keyed on the hostname or domain of the site that set them, that is, their _host key_.   

For example, for cookies from `https://support.chat.example`, the host key is `("support.chat.example")`.  

Under CHIPS, cookies that opt into partitioning will be double-keyed on their host key and _partition key_.   

> A _cookie's partition key_ is the site ([scheme and registrable domain](https://web.dev/same-site-same-origin/#%22schemeful-same-site%22)) of the top-level URL the browser was visiting at the start of the request to the endpoint that set the cookie.  

Reviewer: ```suggestion
A _cookie's partition key_ is the site ([scheme and registrable domain](https://web.dev/same-site-same-origin/#%22schemeful-same-site%22)) of the top-level URL the browser is visiting at the start of a request to an endpoint that sets the cookie. This means the cookie partition key is shared by a site and its embedded content.
```

Reviewer: You define the request's partitioned key, which I think you should still do again, but I believe this benefits from secondary explanation.

In the example above where `https://support.chat.example` is embedded on `https://retail.example`, the top-level URL is `https://retail.example`.  

The partition key in that case is `("https", "retail.example")`.  

Likewise, a _request's partition key_ is the site of the top-level URL the browser is visiting at the start of a request. Browsers must only send a cookie with the `Partitioned` attribute in requests with the same partition key as that cookie.  

Here's what the cookie key in the example above looks like before and after CHIPS.

> {% Img src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/ZZm3G3cCgziUK2eezdiu.png", alt="ALT_TEXT_HERE", width="800", height="204" %}

Reviewer: ```suggestion
{% Img
   src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/ZZm3G3cCgziUK2eezdiu.png",
   alt="Site A and the embedded site C share a partitioned cookie. When not embedded, site C cannot access the partitioned cookie.",
   width="800", height="204"
%}
```

**Before CHIPS**

```text
key=("support.chat.example")
```

**After CHIPS**

```text
key={("https", "retail.example"),
     ("support.chat.example")}
```

#### Security design

> To encourage good security practices, CHIPS proposes that partitioned cookies must be bound to the hostname (and not the registrable domain), and only be set by and sent over secure protocols. To ensure this, partitioned cookies must use the ``__Host-`` prefix.

Reviewer: ```suggestion
To encourage good security practices, CHIPS proposes that partitioned cookies must be bound to the hostname (and not the registrable domain). Further, CHIPS proposes cookies only be set by and sent over secure protocols. To ensure this, partitioned cookies must use the ``__Host-`` prefix.
```

The `__Host-` prefix requires that the cookie be set with `Secure` and `Path=/` and disallows the `Domain` attribute. Since `Domain` cookies can be shared between different third-party subdomains within a partition, disallowing it makes partitioned cookies as close to being [origin-bound](docs/privacy-sandbox/glossary/#origin) as possible; and aligns cookies closer to the [Same-Origin Policy](https://developer.mozilla.org/docs/Web/Security/Same-origin_policy).

> Example: 

Reviewer: nit: You say "for example" elsewhere, and I think that's suited here too

```text
Set-Cookie: __Host-example=34d8g; SameSite=None; Secure; Path=/; Partitioned;
```

### Implementation status

Chrome has published an [Intent to Experiment](https://groups.google.com/a/chromium.org/g/blink-dev/c/_dJFNJpf91U) for CHIPS and is in the process of writing code for the feature. Follow the [Chrome Platform Status entry](https://chromestatus.com/feature/5179189105786880) for progress.

CHIPS is already available behind flags in Chrome 99. Check out the testing instructions and demo on [chromium.org](https://www.chromium.org/updates/chips/). Updated testing and debugging details will be added once the origin trials starts.

To try it out locally, turn on the `chrome://flags/#partitioned-cookies` flag in Chrome Canary or use the `--partitioned-cookies=true` command line flag.

> {% Img src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/Afo08gb3WxuNT77Bi6xc.png", alt="ALT_TEXT_HERE", width="779", height="329" %}

Reviewer: ```suggestion
{% Img
   src="image/vgdbNJBYHma2o62ZqYmcnkq3j0o1/Afo08gb3WxuNT77Bi6xc.png",
   alt="Set the partitioned cookies flag to Enabled.",
   width="779", height="329"
%}
```

### Engage and share feedback

+   **GitHub**: Read the [proposal](https://github.com/WICG/CHIPS), [raise questions and follow the discussion](https://github.com/WICG/CHIPS/issues).

> +   **Developer support**: Ask questions and join discussions on the [Privacy Sandbox Developer Support repo](https://github.com/GoogleChromeLabs/privacy-sandbox-dev-support).

Reviewer: ```suggestion
+   **Developer support**: Ask questions and join discussions on the [Privacy Sandbox Developer Support repo](https://github.com/GoogleChromeLabs/privacy-sandbox-dev-support).

```

Reviewer: It's best practice to leave a blank line at the end of markdown files.

Author: I think the linter used to catch that...something weird's going on with the linter lately :)

Cookies Having Independent Partitioned State (CHIPS) is a Privacy Sandbox proposal that will allow developers to opt a cookie into "partitioned" storage, with separate cookie jars per top-level site.

> A partitioned third-party cookie is tied to the top-level site where it's initially set and cannot be accessed by other sites. The aim is to still allow cookies to be set by a third-party service, but those cookies can only be read within the context of the top-level site. 

Reviewer: FYI @mihajlija I committed this change, but you may need to undo.

Author: The order swap works well, wording edits change the meaning so I'll revert that.
